zproject/backends.py

```diff
@@ -1470,6 +1483,133 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+        we will store the parameters in redis under a key derived from the state. That will allow us
+        to validate the state and retrieve the params after the redirect - by querying redis for
+        the key derived from the state sent in the POST redirect.
+        '''
+        request_data = self.strategy.request_data().dict()
+        data_to_store = {
+            key: request_data[key] for key in self.standard_relay_params
+            if key in request_data
+        }
+        state = self.state_token()
+        put_dict_in_redis(redis_client, 'apple_auth_{token}',
+                          data_to_store, self.REDIS_EXPIRATION_SECONDS,
+                          token=state)
+        return state
+
+    def validate_state(self) -> Optional[str]:
+        """
+        Oauth2 state validation, done when the user is redirected to us back from the IdP.
+        This overrides the upstream method, because we need
+        to retrieve certain parameters for this auth attempt that we stored in the first step
+        of the flow under a key derived from the state.
+        """
+        request_state = self.get_request_state()
+
+        if not request_state:
+            logging.info("Sign in with Apple failed: missing state parameter.")
+            raise AuthMissingParameter(self, 'state')
+
+        formatted_request_state = "apple_auth_" + request_state
+        redis_data = get_dict_from_redis(redis_client, "apple_auth_{token}",
+                                         formatted_request_state)
+        if redis_data is None:
+            logging.info("Sign in with Apple failed: bad state token.")
+            raise AuthStateForbidden(self)
+
+        for param, value in redis_data.items():
+            if param in self.standard_relay_params:
+                self.strategy.session_set(param, value)
+        return request_state
+
+    def is_native_flow(self) -> bool:
+        return self.strategy.request_data().get('native_flow', False)
+
+    def decode_id_token(self, id_token: str) -> Dict[str, Any]:
+        '''
+        Decode and validate JWT token from apple and return payload including user data.
+        We override this, because upstream implementation does not support the use of "Bundle ID"
+        which should be used as audience to decode id_token in the iOS native auth flow.
+        We also take advantage of this override to improve error handling and exceptions.
+        '''
+        if self.is_native_flow():
+            audience = self.setting("BUNDLE_ID")
+        else:
+            audience = self.setting("SERVICES_ID")
+
+        try:
+            kid = jwt.get_unverified_header(id_token).get('kid')
+            public_key = RSAAlgorithm.from_jwk(self.get_apple_jwk(kid))
+            decoded = jwt.decode(id_token, key=public_key,
+                                 audience=audience, algorithm="RS256")
+        except PyJWTError:
+            raise AuthFailed(self, "Token validation failed")
```

Comment:
What is the difference between `AuthCanceled` and `AuthFailed`?  That seems to be a change we made here.

For functions like this that we've forked just one line of, we should have a clear comment explaining what we changed and why.  (We may want to submit upstream PRs for those changes to help clean up our implementation; e.g. I would think we'd try to submit upstream a change for this one to accept an optional `audience` argument).

Reply:
From upstream:
```
class AuthCanceled(AuthException):
    """Auth process was canceled by user."""
```
```
class AuthFailed(AuthException):
    """Auth process failed for some reason."""
```

Clearly should be Authfailed here. On top of that, upstream does `raise AuthCanceled("Token validation failed")` which is wrong, because the first argument is the backend, the second arg is the message. Upstream exceptions from here convert to the empty string when doing `str(e)` because of that. I didn't submit a PR upstream, because the maintainer was gone for a couple of months again - but seems to have returned recently, so we should submit these fixes.

They also merged this https://github.com/python-social-auth/social-core/commit/7d3daf501a16cb152298617fa73ecbbf4e129414 which allows passing an `AUDIENCE` list of allowed audiences. But I'm not sure if we like that, because we discussed with Greg a bit in https://chat.zulip.org/#narrow/stream/107-kandra/topic/Apple.20auth/near/886762 and at that point decided to be more strict in the validation - BUNDLE_ID in native flow, SERVICE_id in browser flow. So we might want to keep our custom code here anyway...

Reply:
Or submit a PR upstream to make it better.

Reply:
Yeah. I submitted the error handling fixes for now at https://github.com/python-social-auth/social-core/pull/476

Reply:
OK.  Maybe worth also adding a commit to make `audience` an option argument, which would let us more fully unfork this?